[docs] Revise "Component theming" and "How to customize" guides (#31997)
mui/material-ui: docs/data/material/customization/how-to-customize/how-to-customize.md

# How to customize

> <p class="description">Learn the many options for customizing the Material UI components.</p>

Reviewer: ```suggestion
<p class="description">Learn the ways to customize Material UI components.</p>
```
I feel that `the many options` sounds not common.

There are a few approaches you could choose for cusotmizing a given component styles. It will ultimately depend on your specific context to decide which is the most appropriate one. Let's go through them from the narrowest use case to the broadest:

1. [One-off customization](#1-one-off-customization)
1. [Reusable style overrides](#2-reusable-style-overrides)
1. [Global theme variation](#4-global-theme-variation)
1. [Global CSS override](#5-global-css-override)

## 1. One-off customization

If you want to change the styles of a component in a specific location and don't want it to be perpetuated to other instances of it, you have the following solutions available:

### Use the `sx` prop

The easiest way to add style overrides for a one-off situation is to use the [`sx` prop](/system/basics/#the-sx-prop), which is available on all Material UI components.
Here is an example:

{{"demo": "SxProp.js" }}

#### Overriding nested component styles

> If you want to customize very specific parts of a component, you can use classes provided by Material UI inside the `sx` prop. As an example, let's say you want to change the Slider's thumb from a circle to a square.

Reviewer 2: ```suggestion
If you want to customize a specific part of a component, you can use the class name provided by Material UI inside the `sx` prop. 
	
	As an example, let's say you want to change the `Slider` component's thumb from a circle to a square.
```

One step further :)

First step would be using the browser dev tools to identify the class for the component slot you want to override:

<img src="/static/images/customization/dev-tools.png" alt="dev-tools" width="796" style="margin-bottom: 16px;" />

Then, copy the target class of `MuiSlider-thumb`, place it inside the styles witihn the `sx` prop, and add your overrides.

{{"demo": "DevTools.js"}}

The styles injected into the DOM by Material UI rely on class names that [follow a simple pattern](/styles/advanced/#class-names):
`[hash]-Mui[Component name]-[name of the slot]`.

In this case, the styles are applied with `.css-ae2u5c-MuiSlider-thumb` but you only really need to target the `.MuiSlider-thumb`, where `Slider` is the component and `thumb` is the slot.

> ⚠️ Note that these class names can't be used as CSS selectors because they are unstable.

### Overriding styles with class names

If you want to override the component styles using custom classes, you can use the `className` prop, available on each component. Use the same approach described above if you want to override specific parts of the component.

> Go to the [Styles Library Interoperability](/guides/interoperability/) guide to find examples of this approach using different styling libraries.

Author: Looking forward to that one!

# How to customize

<p class="description">Learn how to customize Material UI components by taking advantage of different strategies for specific use cases.</p>

Material UI provides several different ways to customize a component's styles. Your specific context will determine which one is ideal. From narrowest to broadest use case, here are the options:

1. [One-off customization](#1-one-off-customization)
1. [Reusable style overrides](#2-reusable-style-overrides)
1. [Global theme variation](#4-global-theme-variation)
1. [Global CSS override](#5-global-css-override)

## 1. One-off customization

To change the styles of _one single instance_ of a component, you can use one of the following options:

### The `sx` prop

> The easiest way to add style overrides for a one-off situation is to use the [`sx` prop](/system/basics/#the-sx-prop), which is available on all Material UI components.

Reviewer 3: Let's rephrase this to remove the word "easiest" - could be discouraging to someone who doesn't agree that it's easy. 😛

```suggestion
The [`sx` prop](/system/basics/#the-sx-prop) is the best option for adding style overrides to a single instance of a component in most cases.
It can be used with all Material UI components.
```

It can be used with all Material UI components.

{{"demo": "SxProp.js" }}

#### Overriding nested component styles

To customize a specific part of a component, you can use the class name provided by Material UI inside the `sx` prop.

As an example, let's say you want to change the `Slider` component's thumb from a circle to a square.

> First, use your browser's dev tools to identify the class for the component slot you want to override:

Reviewer 4: I think that it was clearer before. We were explaining how to "identify the class". How will people know that `MuiSlider-thumb` is the right selector when all they see is?

<img width="338" alt="Screenshot 2022-04-07 at 12 55 19" src="https://user-images.githubusercontent.com/3165635/162183748-96a94722-e1af-40d6-b020-29dc1a2dc04a.png">

We do no longer explain it. It's not trivial.

Author: You're right. Revisited this one now, let me know your thoughts.

> <img src="/static/images/customization/dev-tools.png" alt="dev-tools" width="796" style="margin-bottom: 16px;" />

Reviewer 4: The change of docs/public/static/images/customization/dev-tools.png looks blurry on my screen. Am I the only one? I mean:

a. https://mui.com/customization/how-to-customize/#overriding-nested-component-styles
b. https://deploy-preview-31997--material-ui.netlify.app/customization/how-to-customize/#the-sx-prop

a. looks better on my end. I'm raising this because it's not the first time I see this problem. What do you think of limiting the resolution to x2? Otherwise, could we try a fixed integer increment? Currently, we are doing x3.28, maybe x3 or x4 would look crisp?

Author: Huh... looks a bit blurry to me too, now that I'm looking at it again. But I did export at 2x (don't know where you got x3.28 from!). Though the blurriness could be because I increased the image size a bit. Let me see if I can tweak it properly.

> Then, copy the target class of `MuiSlider-thumb`, place it inside the styles witihn the `sx` prop, and add your overrides.

Reviewer 4: This reads strange, it doesn't feel like it's following the standard CSS terminology.

```suggestion
Then, use this class name (`MuiSlider-thumb`) to write a CSS selector within the `sx` prop (`& .MuiSlider-thumb`), and add your overrides inside.
```

{{"demo": "DevTools.js"}}

> The styles injected into the DOM by Material UI rely on class names that [follow a simple pattern](/styles/advanced/#class-names):

Reviewer 3: Let's replace "simple" here for the same reason as "easy" above.

```suggestion
The styles injected into the DOM by Material UI rely on class names that all [follow a standard pattern](/styles/advanced/#class-names):
```

<MenuItem selected className="MenuItem">
```

> If you'd like to learn more about CSS specificity, we recommend checking [MDN's guide](https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity) about it.

Reviewer 3: Let's get more... specific 😂

```suggestion
If you'd like to learn more about this topic, we recommend checking out [the MDN Web Docs on CSS Specificity](https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity).
```

#### Why do I need to increase specificity to override one component state?

By design, the CSS specification makes the pseudo-classes increase the specificity.
For consistency with native elements, Material UI increases the specificity of its custom state classes.

> This has one important advantage: it allows you to cherry-pick the state you want to customize.

Reviewer 3: I don't think I fully understand the answer to this question. Here's what I gather:

- pseudo-classes have a high level of specificity
- MUI's state classes also have a high level of specificity - possibly more than CSS pseudo-classes?
- this is a good thing because it makes it simple to target individual components using these classes

Is that correct?

Reviewer 2: - pseudo-classes have a high level of specificity ✅ 
- MUI's state classes also have a high level of specificity ~possibly more than CSS pseudo-classes~? they have the same as pseudo states, see https://codepen.io/mnajdova/pen/mdpwxyY
- this is a good thing because it makes it simple to target individual ~components~ component's state using these classes

<MenuItem selected className="MenuItem">
```

If you'd like to learn more about this topic, we recommend checking out [the MDN Web Docs on CSS Specificity](https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity).

#### Why do I need to increase specificity to override one component state?

> By design, the CSS specification makes the pseudo-classes increase the specificity.

Reviewer 3: ```suggestion
CSS pseudo-classes have a high level of specificity.
```

> For consistency with native elements, Material UI increases the specificity of its custom state classes.
> This has one important advantage: it allows you to cherry-pick the state you want to customize.

Reviewer 3: Following up on this - I think this is worthwhile to add for clarity.

```suggestion
For consistency with native elements, Material UI's state classes have the same level of specificity as CSS pseudo-classes, making it possible to target an individual component's state.
```

| required      | `.Mui-required`     |
| selected      | `.Mui-selected`     |

> > ⚠️ Never style these state classes' names directly. If you do that, it will impact all the components with unclear side-effects. Always target it together with a component.

Reviewer 3: Rephrasing to avoid the awkward possessive.

```suggestion
> ⚠️ Never apply styles directly to state class names. This will impact all components with unclear side-effects. Always target a state class together with a component.
```

> ## 2. Reusable style overrides

Reviewer: Is it better to say `2. Reusable component`?

Author: Probably yes! Sounds better.

> If you find that you need to reuse the same overrides in different locations across your application, you can use the [`styled()`](/system/styled/) utility to create a reusable component:

Reviewer 3: Trimming.

```suggestion
To reuse the same overrides in different locations across your application, create a reusable component using the [`styled()`](/system/styled/) utility:
```

{{"demo": "StyledCustomization.js", "defaultCodeOpen": true}}

### Dynamic overrides

> Using the `styled()` utility above allows you to add dynamic styles based on the component's props. Here are two ways you can do that:

Reviewer 3: Let's state the options right here.

```suggestion
The `styled()` utility lets you add dynamic styles based on a component's props.
You can do this with **dynamic CSS** or **CSS variables**.

```

#### Dynamic CSS

> > ⚠️ Note that if you are using TypeScript you will need to update the prop's types of the new component.

Reviewer 3: ```suggestion
> ⚠️ Note that if you are using TypeScript you will need to update the prop types of the new component.
```

Reviewer 2: I feel like we should move this back right before the demo showing how it should be altered in TypeScript.

Author: Before or after? I haven't altered its position.

}));
```

#### CSS variables

{{"demo": "DynamicCSSVariables.js"}}

## 3. Global theme overrides

> If you want to secure styling consistency between components, managing your user interface as a whole, Material UI provides a way to customize the components from the theme level. We have a dedicated guide about this approach, take a look at the [Component theming customization](/customization/theme-components/) page for more details.

Reviewer 3: Trimming.

```suggestion
Material UI provides theme tools for managing style consistency between all components across your user interface. 
Visit the [Component theming customization](/customization/theme-components/) page for more details.
```

## 4. Global CSS override

> If you just want to add some global baseline styles for some of the HTML elements, you can use the `GlobalStyles` component. Here is an example of how you can override styles for the `h1` elements.

Reviewer 3: ```suggestion
To add global baseline styles for some of the HTML elements, use the `GlobalStyles` component.

Here is an example of how you can override styles for the `h1` elements:
```
